File: problems/matrixChain/cpp/matrix_chain.cpp
// File: mcm_cpp.cpp
#include <iostream>
#include <vector>
#include <future>
#include <limits>
#include <numeric> // for std::iota (optional)
#include <algorithm> // for std::min
#include <stdexcept> // for std::runtime_error
#include <chrono>    // For timing (optional)

// Helper function to compute the cost for a single entry m[i][j]
// It reads previously computed values from the shared 'm' table.
int compute_entry(int i, int j, const std::vector<int>& p, const std::vector<std::vector<int>>& m) {
    long long min_cost_ll = std::numeric_limits<long long>::max();
    int n = p.size() - 1; // Needed for dimension access p[k]

    if (i < 1 || i > n || j < i || j > n) {
         throw std::runtime_error("Invalid indices in compute_entry");
    }

    for (int k = i; k < j; ++k) {
        // Check indices for 'm' access
        if (k < i || k+1 > j || k+1 < 1) {
             throw std::runtime_error("Invalid k index");
        }
        // Check indices for 'p' access
        if (i - 1 < 0 || i - 1 >= p.size() || k < 0 || k >= p.size() || j < 0 || j >= p.size()) {
             throw std::runtime_error("Invalid dimension index");
        }

        // Ensure m values are non-negative (or handle sentinel if used)
        if (m[i][k] < 0 || m[k + 1][j] < 0) {
            // This shouldn't happen if max_int was used correctly for uncomputed/infinite
             if (m[i][k] == std::numeric_limits<int>::max() || m[k + 1][j] == std::numeric_limits<int>::max()) {
                continue; // Skip if subproblem cost is infinite
             }
            throw std::runtime_error("Negative cost encountered in subproblem");
        }

        // Use long long for intermediate calculation to prevent overflow
        long long term1 = m[i][k];
        long long term2 = m[k + 1][j];
        long long term3 = static_cast<long long>(p[i - 1]) * p[k] * p[j];

        // Check for potential overflow before adding
        if (term1 > std::numeric_limits<long long>::max() - term2) {
             // Overflow would occur on term1 + term2
             continue; // Skip this k value
        }
        long long sum12 = term1 + term2;
         if (term3 > std::numeric_limits<long long>::max() - sum12) {
             // Overflow would occur on sum12 + term3
             continue; // Skip this k value
         }

        long long current_cost_ll = sum12 + term3;
        min_cost_ll = std::min(min_cost_ll, current_cost_ll);
    }

    // Check if the result fits in int
    if (min_cost_ll > std::numeric_limits<int>::max()) {
        return std::numeric_limits<int>::max();
    }
    // Handle case where no valid 'k' was found (should only happen if subproblems were max_int)
    if (min_cost_ll == std::numeric_limits<long long>::max()) {
         return std::numeric_limits<int>::max();
    }

    return static_cast<int>(min_cost_ll);
}

// Parallel MCM using std::async and std::future
int matrix_chain_order_async(const std::vector<int>& p) {
    int n = p.size() - 1;
    if (n <= 0) return 0;

    // DP table: m[i][j] stores the min cost for matrices Ai...Aj
    // Initialize with 0. m[i][i] = 0 is correct. Other values will be overwritten.
    std::vector<std::vector<int>> m(n + 1, std::vector<int>(n + 1, 0));

    // Iterate over chain lengths
    for (int len = 2; len <= n; ++len) {
        std::vector<std::future<int>> futures;
        // Store indices to update 'm' later, corresponding to futures
        std::vector<std::pair<int, int>> indices;

        // Launch tasks for all subproblems of the current length 'len'
        for (int i = 1; i <= n - len + 1; ++i) {
            int j = i + len - 1;
            indices.emplace_back(i, j); // Store the target indices

            // Launch asynchronously. Pass 'm' by const reference because
            // compute_entry only needs read access to *previously* computed values.
            // Use std::launch::async policy to suggest thread creation.
            futures.push_back(std::async(std::launch::async, compute_entry, i, j, std::cref(p), std::cref(m)));
        }

        // Wait for all tasks for the current length and collect results
        for (size_t k = 0; k < futures.size(); ++k) {
            try {
                int result = futures[k].get(); // Waits for the future, throws if task threw exception
                int i = indices[k].first;
                int j = indices[k].second;
                m[i][j] = result; // Update the DP table
            } catch (const std::exception& e) {
                std::cerr << "Error computing m[" << indices[k].first << "][" << indices[k].second << "]: " << e.what() << std::endl;
                // Decide how to handle errors, e.g., rethrow, return error code, etc.
                // Setting to max_int might be an option if partial results are acceptable.
                 m[indices[k].first][indices[k].second] = std::numeric_limits<int>::max();
            }
        }
         // At this point, all m[i][j] for the current 'len' are computed.
    }

    return m[1][n];
}


// Standard Sequential MCM (for comparison)
int matrix_chain_order_seq(const std::vector<int>& p) {
    int n = p.size() - 1;
    if (n <= 0) return 0;
    std::vector<std::vector<int>> m(n + 1, std::vector<int>(n + 1, 0));

    for (int len = 2; len <= n; ++len) {
        for (int i = 1; i <= n - len + 1; ++i) {
            int j = i + len - 1;
            m[i][j] = std::numeric_limits<int>::max(); // Initialize with infinity
            for (int k = i; k < j; ++k) {
                 // Use long long for intermediate calculation
                 long long cost_ll = static_cast<long long>(m[i][k])
                                    + m[k + 1][j]
                                    + static_cast<long long>(p[i - 1]) * p[k] * p[j];

                 // Clamp to int max if overflow occurs
                 int cost = (cost_ll > std::numeric_limits<int>::max())
                            ? std::numeric_limits<int>::max()
                            : static_cast<int>(cost_ll);

                 if (cost < m[i][j]) {
                    m[i][j] = cost;
                 }
            }
        }
    }
    return m[1][n];
}

// int main() {
//     std::vector<int> p = {30, 35, 15, 5, 10, 20, 25}; // Example dimensions

//     std::cout << "Dimensions: ";
//     for (size_t i = 0; i < p.size(); ++i) {
//         std::cout << p[i] << (i == p.size() - 1 ? "" : " x ");
//     }
//     std::cout << std::endl;

//     try {
//         // Time sequential version (optional)
//         auto start_seq = std::chrono::high_resolution_clock::now();
//         int cost_seq = matrix_chain_order_seq(p);
//         auto end_seq = std::chrono::high_resolution_clock::now();
//         std::chrono::duration<double, std::milli> duration_seq = end_seq - start_seq;
//         std::cout << "Sequential Min Cost: " << cost_seq << " (took " << duration_seq.count() << " ms)" << std::endl;


//         // Time parallel version
//         auto start_async = std::chrono::high_resolution_clock::now();
//         int cost_async = matrix_chain_order_async(p);
//         auto end_async = std::chrono::high_resolution_clock::now();
//         std::chrono::duration<double, std::milli> duration_async = end_async - start_async;
//         std::cout << "Async Min Cost:      " << cost_async << " (took " << duration_async.count() << " ms)" << std::endl;

//         if(cost_seq != cost_async) {
//             std::cerr << "Warning: Sequential and Async results differ!" << std::endl;
//         }

//     } catch (const std::exception& e) {
//         std::cerr << "An error occurred: " << e.what() << std::endl;
//         return 1;
//     }

//     return 0;
// }

#include <random> // Include this at the top with other headers!

int main() {
    const int num_matrices = 200; // You can increase this to 500, 1000, etc.
    const int min_dim = 5;         // Minimum dimension size
    const int max_dim = 100;       // Maximum dimension size

    std::vector<int> p(num_matrices + 1); // Need n+1 dimensions for n matrices

    // Initialize random number generator
    std::random_device rd;
    std::mt19937 gen(rd()); // Mersenne Twister RNG
    std::uniform_int_distribution<> dis(min_dim, max_dim);

    for (int i = 0; i <= num_matrices; ++i) {
        p[i] = dis(gen);
    }

    std::cout << "Number of matrices: " << num_matrices << std::endl;
    std::cout << "Sample dimensions (first 10): ";
    for (int i = 0; i < std::min(10, (int)p.size()); ++i) {
        std::cout << p[i] << (i == std::min(10, (int)p.size()) - 1 ? "" : " x ");
    }
    if (p.size() > 10) std::cout << " ...";
    std::cout << std::endl;

    try {
        // Time sequential version
        auto start_seq = std::chrono::high_resolution_clock::now();
        int cost_seq = matrix_chain_order_seq(p);
        auto end_seq = std::chrono::high_resolution_clock::now();
        std::chrono::duration<double, std::milli> duration_seq = end_seq - start_seq;
        std::cout << "Sequential Min Cost: " << cost_seq << " (took " << duration_seq.count() << " ms)" << std::endl;

        // Time parallel version
        auto start_async = std::chrono::high_resolution_clock::now();
        int cost_async = matrix_chain_order_async(p);
        auto end_async = std::chrono::high_resolution_clock::now();
        std::chrono::duration<double, std::milli> duration_async = end_async - start_async;
        std::cout << "Async Min Cost:      " << cost_async << " (took " << duration_async.count() << " ms)" << std::endl;

        if (cost_seq != cost_async) {
            std::cerr << "Warning: Sequential and Async results differ!" << std::endl;
        }

    } catch (const std::exception& e) {
        std::cerr << "An error occurred: " << e.what() << std::endl;
        return 1;
    }

    return 0;
}

File: problems/matrixChain/ocaml/matrix_chain.ml
open Lwt.Infix
(* open Unix *)

(* Standard Sequential MCM *)
let matrix_chain_order_seq p =
  let n = Array.length p - 1 in
  if n <= 0 then 0
  else
    let m = Array.make_matrix (n + 1) (n + 1) 0 in
    for len = 2 to n do
      for i = 1 to n - len + 1 do
        let j = i + len - 1 in
        m.(i).(j) <- max_int;
        for k = i to j - 1 do
          let cost = Int64.(
            add (add (of_int m.(i).(k)) (of_int m.(k+1).(j)))
                (mul (mul (of_int p.(i-1)) (of_int p.(k))) (of_int p.(j)))
          ) in
          let int_cost =
            if cost > Int64.of_int max_int then max_int
            else Int64.to_int cost
          in
          if int_cost < m.(i).(j) then
            m.(i).(j) <- int_cost
        done
      done
    done;
    m.(1).(n)

(* Parallel MCM using Lwt *)
let matrix_chain_order_lwt p =
  let n = Array.length p - 1 in
  if n <= 0 then Lwt.return 0
  else
    let m = Array.make_matrix (n + 1) (n + 1) 0 in

    let compute_entry i j =
      let min_cost = ref max_int in
      for k = i to j - 1 do
        let cost = Int64.(
          add (add (of_int m.(i).(k)) (of_int m.(k+1).(j)))
              (mul (mul (of_int p.(i-1)) (of_int p.(k))) (of_int p.(j)))
        ) in
        let int_cost =
          if cost > Int64.of_int max_int then max_int
          else Int64.to_int cost
        in
        if int_cost < !min_cost then
          min_cost := int_cost
      done;
      m.(i).(j) <- !min_cost;
      Lwt.return ()
    in

    let rec loop len =
      if len > n then
        Lwt.return m.(1).(n)
      else
        let tasks = ref [] in
        for i = 1 to n - len + 1 do
          let j = i + len - 1 in
          tasks := compute_entry i j :: !tasks
        done;
        Lwt.join !tasks >>= fun () ->
        loop (len + 1)
    in
    loop 2

(* Helper to generate random dimensions *)
let generate_random_dimensions n ~min_dim ~max_dim seed =
  Random.init seed;
  Array.init (n + 1) (fun _ -> min_dim + Random.int (max_dim - min_dim + 1))

(* Timing helper *)
let time f =
  let t1 = Unix.gettimeofday () in
  let res = f () in
  let t2 = Unix.gettimeofday () in
  Printf.printf "Time: %.4fs\n" (t2 -. t1);
  res

(* Async timing helper *)
let time_lwt f =
  let t1 = Unix.gettimeofday () in
  f () >>= fun res ->
  let t2 = Unix.gettimeofday () in
  Printf.printf "Time: %.4fs\n" (t2 -. t1);
  Lwt.return res

(* Example Usage *)
let () =
  let num_matrices = 200 in
  let seed = 42 in
  let p = generate_random_dimensions num_matrices ~min_dim:5 ~max_dim:100 seed in
  Printf.printf "Generated %d matrices (dimensions array size: %d)\n"
    num_matrices (Array.length p);

  Printf.printf "Running sequential version...\n";
  let cost_seq = time (fun () -> matrix_chain_order_seq p) in
  Printf.printf "Sequential Min Cost: %d\n\n" cost_seq;

  Printf.printf "Running parallel (Lwt) version...\n";
  let cost_lwt = Lwt_main.run (
    time_lwt (fun () -> matrix_chain_order_lwt p)
  ) in
  Printf.printf "Lwt Min Cost: %d\n" cost_lwt
